src/ssort/_requirements.py

```diff
@@ -23,15 +25,11 @@ class Requirement:
     scope: Scope = Scope.LOCAL
 
 
-@functools.singledispatch
-def get_requirements(node):
-    raise NotImplementedError(
-        f"could not find requirements for unsupported node:  {node!r}"
-        f"at line {node.lineno}, column: {node.col_offset}"
-    )
+_requirements_visitor: NodeVisitor[Requirement] = NodeVisitor()
+get_requirements = _requirements_visitor.visit
```

Comment:
This is an improvement, but I'm concerned that there are now two modes of recursion:
  1. A `_get_requirements` function recurses by calling `get_requirements`.
  2. A visitor function calls `visitor.visit`.

I think there are two options:
  1. Move recursion entirely into the visitor:  Allow get bindings to yield only the bindings that are bound directly by the visited node.
  2. Move recursion entirely into `get_requirements`:  Make the visitor just yield immediate child nodes and `for child in visit(node): yield from get_requirements(child)` as the default implementation.

My vote is with 2, as I think people will always want to read the get_requirements/get_binding functions first, and it makes it a bit easier to stop recursion if necessary.

Reply:
Perhaps the other way to frame this is in terms of composition vs inheritance.  Effectively a NodeVisitor instance is a kind of class that subclasses the basic implementation.  I'm struggling to come up with a coherent argument from there, but that's the instinct that this is triggering.

Reply:
Hmm, that does seem a lot cleaner. Perhaps what we want is to re-purpose the `_visitor` module to provide a high speed `iter_child_nodes` function which parallels the `ast.iter_child_node` function. I think that would enforce a cleaner separation between the `_visitor` module and the `_bindings` and `_requirements` modules.